### DISCUSSION TITLE
Vercel calling my APIs twice during deploys

### DISCUSSION THREAD

**Discussion by JamesSingleton** [vercel/community#3038] [Help] [answered]
### Summary

I use Sanity for my CMS and started to notice that my API usage was through the roof on a website that maybe 20-30 people know about. I started to look over the logs and noticed that during builds/deploys my routes are getting hit twice and therefore (presumably), hitting the API twice. I have about 305 static pages and if I do one deploy, that is already 610 API calls to Sanity. This gets even worse if I make multiple deploys in a day

![image](https://github.com/vercel/community/assets/21000200/5e4d226f-3113-4151-82ae-9edd75aff71b)

### Steps to Reproduce

Have a app directory project setup to call Sanity and `generateStaticParams()` from Sanity and push your changes so that Vercel deploys them... I honestly don't know, I am just here because I can't contact Vercel Support directly due to being on the Hobby Tier.

**Comment by amyegan** [comment]
Hi @JamesSingleton. I have a Sanity Next.js test app and I'm seeing the expected number of API calls in my Sanity usage after making a new deployment. But my project only has a couple of pages since it's only a test site. 

If you share your public repo or a minimal reproduction, that will let me get a better look at what is happening for your project.

**Reply by smaeda-ks** [reply, marked as answer]
@JamesSingleton Just to be clear, those logs are requests to your static pages that were already generated during the deployment process, and NOT causing any additional API requests to your CMS backend. In fact, if you look at the user-agent closely, it says "Vercelbot", which means it's coming from our "screenshot" functionality, where you can see the screenshot of your deployed page on each deployment on the UI.

So it's just the fact that our bot is accessing your deployment upon successful deployment and fetching pages (i.e., [prefetching](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#prefetching)) to simulate a request from the browser. And why you see those duplicate requests is that the bot is first sending a Range request to fetch the minimal quantity of body bytes just to check the site reachability. After this, it requests the whole body to fetch the actual content. This is why you see two different status codes 200/206, and they're expected.

**Reply by JamesSingleton** [reply]
@smaeda-ks Interesting 🤔 I guess time to investigate why a website with minimal traffic is all of a sudden getting 115k API requests to Sanity. It was just odd that I was seeing multiple requests to the same route in the logs during the deploy process (or the 30 min window around a deploy). One with `got` as the UA and one as `VercelBot`. Also, usually the home page is the screenshot of the deploy, so interesting to see it takes a screenshot of every single route, learned something new.

**Reply by smaeda-ks** [reply]
> Also, usually the home page is the screenshot of the deploy, so interesting to see it takes a screenshot of every single route, learned something new.

As I mentioned, this is most likely the effect of prefetching:
https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#prefetching

Not a bad thing at all, to be fair.

**Reply by JamesSingleton** [reply]
The reason I originally came here is because if you look at the spike I hovered over for the 28th (not entirely sure if this is UTC or local June 28th) is when I made 7 deploys. And the days the have less than 10k I have maybe 1 or 2 deploys and the rest is what I would assume from traffic. Furthermore, if you look at June 3rd (if this is UTC it would I guess count June 2nd  for me if I am doing my math correctly) you will see another spike, I had 12 deploys. June 3rd was over 30k requests to Sanity's API which is not something I would assume from traffic.

So for me it is a very big coincidence that on the days I have a lot of deploys are the days I have spikes in my API hits. I would also like to keep prefetch if possible because without it, transitioning from one page to another is SLOW.

<img width="867" alt="Screenshot 2023-06-28 at 6 07 05 PM" src="https://github.com/vercel/next.js/assets/21000200/56038ac8-c147-4f05-b760-7f96452692ef">

**Reply by smaeda-ks** [reply]
Again, what you see in your Vercel logs are requests to your "static" == pre-rendered pages, which do not trigger any external calls to Sanity (it's a post-build stage). In your case, those only happen during the "build" time. Those Vercel logs are irrelevant to your main issue. Disabling prefetch changes nothing.

You should ask Sanity for more verbose logs for those API calls to see how exactly they're generated. We have no visibility into the outgoing request logs from your past build environment, as they're ephemeral.

**Reply by JamesSingleton** [reply]
So if I wasn't calling `generateStaticParams()` would those calls in the Vercel logs be going to the static pre-rendered pages at that point?  I just implemented `generateStaticParams()` in the last day or two. Just trying to rule some things out.

**Reply by JamesSingleton** [reply]
Also, looking at the logs again tonight, seeing VercelBot hitting some URLs but I didn't make any deployment at this time. I am guessing this is due to the prefetching and it is prefetching all the categories that are visible on the page. But they _shouldn't_ be making any API calls right, just the static pre-rendered page.

<img width="2536" alt="Screenshot 2023-06-28 at 9 14 47 PM" src="https://github.com/vercel/community/assets/21000200/24f78a5c-1a63-4a3c-a33d-d7738e9157c6">

**Reply by smaeda-ks** [reply]
> Also, looking at the logs again tonight, seeing VercelBot hitting some URLs but I didn't make any deployment at this time. I am guessing this is due to the prefetching and it is prefetching all the categories that are visible on the page. But they shouldn't be making any API calls right, just the static pre-rendered page.

Yes, if the previous screenshot cache is gone at the time when you access the individual deployment page on the UI, the bot will fetch the top page again, and you will see that. And if you're using `generateStaticParams()` and generating statically generated routes, those pages are pre-rendered during build time, and visiting those pages afterward will not trigger API calls to your backend. That's only happening during build time, as I mentioned earlier.
